File: go/navigate.go
package struple

// Navigation / query over an encoded struple buffer.
//
// A buffer is a stream of elements; these helpers slice and inspect it without
// decoding values. Everything is zero-copy — results are sub-slices of the
// input, each itself a valid struple buffer, so it all composes and recurses.
//
// Stream ops (Count, At, Head, Tail, NthRest, Take) operate on the buffer's
// top-level element stream. To descend into an array/map/set, use ContainedItems
// to get its inner stream, then view that.

// View is a zero-copy inspector over an encoded struple buffer.
type View struct {
	Bytes []byte
}

// NewView returns a View over bytes.
func NewView(bytes []byte) View { return View{Bytes: bytes} }

func (v View) reader() *Reader { return NewReader(v.Bytes) }

// Count returns the number of top-level elements.
func (v View) Count() (int, error) {
	r := v.reader()
	n := 0
	for {
		ok, err := r.Skip()
		if err != nil {
			return 0, err
		}
		if !ok {
			break
		}
		n++
	}
	return n, nil
}

// At returns the element at index as a zero-copy sub-view. ok is false if out of
// range.
func (v View) At(index int) (view []byte, ok bool, err error) {
	r := v.reader()
	for i := 0; ; i++ {
		view, ok, err = r.NextView()
		if err != nil || !ok {
			return nil, false, err
		}
		if i == index {
			return view, true, nil
		}
	}
}

// Head returns the first element. ok is false if empty.
func (v View) Head() (view []byte, ok bool, err error) { return v.At(0) }

// Tail returns everything after the first element (empty if 0 or 1 elements).
func (v View) Tail() ([]byte, error) {
	r := v.reader()
	if _, _, err := r.NextView(); err != nil {
		return nil, err
	}
	return r.Rest(), nil
}

// NthRest drops n elements and returns the remaining stream.
func (v View) NthRest(n int) ([]byte, error) {
	r := v.reader()
	for i := 0; i < n; i++ {
		ok, err := r.Skip()
		if err != nil {
			return nil, err
		}
		if !ok {
			break
		}
	}
	return r.Rest(), nil
}

// Take returns the first n elements as a contiguous sub-view.
func (v View) Take(n int) ([]byte, error) {
	r := v.reader()
	for i := 0; i < n; i++ {
		ok, err := r.Skip()
		if err != nil {
			return nil, err
		}
		if !ok {
			break
		}
	}
	return v.Bytes[:len(v.Bytes)-len(r.Rest())], nil
}

// HeadType returns the type code of the first element. ok is false if empty.
func (v View) HeadType() (byte, bool) {
	if len(v.Bytes) > 0 {
		return v.Bytes[0], true
	}
	return 0, false
}

func (v View) IsNil() bool       { t, ok := v.HeadType(); return ok && t == tcNil }
func (v View) IsUndefined() bool { t, ok := v.HeadType(); return ok && t == tcUndef }

func (v View) IsBool() bool {
	t, ok := v.HeadType()
	return ok && (t == tcBoolFalse || t == tcBoolTrue)
}

func (v View) IsInt() bool {
	t, ok := v.HeadType()
	if !ok {
		return false
	}
	return t == tcIntZero || t == tcIntNegBig || t == tcIntPosBig ||
		(t >= tcIntNegMin && t <= tcIntNegMax) ||
		(t >= tcIntPosMin && t <= tcIntPosMax)
}

func (v View) IsFloat() bool {
	t, ok := v.HeadType()
	return ok && (t == tcFloat32 || t == tcFloat64)
}

func (v View) IsDecimal() bool { t, ok := v.HeadType(); return ok && t == tcDecimal }

func (v View) IsNumber() bool { return v.IsInt() || v.IsFloat() || v.IsDecimal() }

func (v View) IsTimestamp() bool { t, ok := v.HeadType(); return ok && t == tcTimestamp }
func (v View) IsUUID() bool      { t, ok := v.HeadType(); return ok && t == tcUUID }
func (v View) IsString() bool    { t, ok := v.HeadType(); return ok && t == tcString }
func (v View) IsBytes() bool     { t, ok := v.HeadType(); return ok && t == tcBytes }
func (v View) IsArray() bool     { t, ok := v.HeadType(); return ok && t == tcArray }
func (v View) IsMap() bool       { t, ok := v.HeadType(); return ok && t == tcMap }
func (v View) IsSet() bool       { t, ok := v.HeadType(); return ok && t == tcSet }

func (v View) IsContainer() bool {
	t, ok := v.HeadType()
	return ok && (t == tcArray || t == tcMap || t == tcSet)
}

// ContainerBody returns the first element's framed body when it is a container
// (escapes intact, zero-copy). ok is false if the head isn't a container.
func (v View) ContainerBody() (body []byte, ok bool, err error) {
	if !v.IsContainer() {
		return nil, false, nil
	}
	r := v.reader()
	e, has, err := r.Next()
	if err != nil || !has {
		return nil, false, err
	}
	switch e.Kind {
	case KindArray, KindMap, KindSet:
		return e.Body, true, nil
	default:
		return nil, false, nil
	}
}

// ContainedItems returns the container's inner element stream, un-escaped (a
// fresh slice owned by the caller). View it with NewView, or a map with
// NewMapView. ok is false if the head isn't a container.
func (v View) ContainedItems() (inner []byte, ok bool, err error) {
	body, ok, err := v.ContainerBody()
	if err != nil || !ok {
		return nil, ok, err
	}
	return Unescape(body), true, nil
}

// ---------------------------------------------------------------------------
// MapView — reads key/value pairs from a map's inner stream
// ---------------------------------------------------------------------------

// MapEntry is a key/value pair of encoded element views.
type MapEntry struct {
	Key   []byte
	Value []byte
}

// MapView reads key/value pairs from a map's inner stream (the un-escaped body
// from View.ContainedItems). Keys are in canonical (sorted) order, so Get
// early-exits.
type MapView struct {
	Inner []byte
}

// NewMapView returns a MapView over a map's inner stream.
func NewMapView(inner []byte) MapView { return MapView{Inner: inner} }

// Count returns the number of entries.
func (m MapView) Count() (int, error) {
	n, err := NewView(m.Inner).Count()
	return n / 2, err
}

// MapIterator yields entries of a MapView in canonical order.
type MapIterator struct {
	r *Reader
}

// Next returns the next entry. ok is false at end of stream.
func (it *MapIterator) Next() (e MapEntry, ok bool, err error) {
	k, ok, err := it.r.NextView()
	if err != nil || !ok {
		return MapEntry{}, false, err
	}
	val, ok, err := it.r.NextView()
	if err != nil {
		return MapEntry{}, false, err
	}
	if !ok {
		return MapEntry{}, false, ErrTruncated
	}
	return MapEntry{Key: k, Value: val}, true, nil
}

// Iterator returns an iterator over the map's entries in canonical order.
func (m MapView) Iterator() *MapIterator { return &MapIterator{r: NewReader(m.Inner)} }

// Get looks up key (the encoded bytes of a key element). It returns the value's
// encoded bytes; ok is false if absent. Ordered scan with early exit.
func (m MapView) Get(key []byte) (value []byte, ok bool, err error) {
	it := m.Iterator()
	for {
		e, has, err := it.Next()
		if err != nil || !has {
			return nil, false, err
		}
		switch bytesCompare(e.Key, key) {
		case 0:
			return e.Value, true, nil
		case 1:
			return nil, false, nil
		}
	}
}

// Indexed materializes a random-access index for O(log n) Get and O(1) At.
func (m MapView) Indexed() (*IndexedMap, error) { return NewIndexedMap(m.Inner) }

// ---------------------------------------------------------------------------
// IndexedMap — random-access index over a map's entries
// ---------------------------------------------------------------------------

// IndexedMap materializes a map's entries into a random-access index. Building it
// is one O(n) pass over the inner stream; thereafter Get is an O(log n) binary
// search (canonical key order means a key compare is the sort order) and At is
// O(1). The entry slices borrow the inner stream, so keep it alive.
type IndexedMap struct {
	Entries []MapEntry
}

// NewIndexedMap builds an index from a map's inner stream (the un-escaped body
// from View.ContainedItems).
func NewIndexedMap(inner []byte) (*IndexedMap, error) {
	var entries []MapEntry
	r := NewReader(inner)
	for {
		k, ok, err := r.NextView()
		if err != nil {
			return nil, err
		}
		if !ok {
			break
		}
		val, ok, err := r.NextView()
		if err != nil {
			return nil, err
		}
		if !ok {
			return nil, ErrTruncated
		}
		entries = append(entries, MapEntry{Key: k, Value: val})
	}
	return &IndexedMap{Entries: entries}, nil
}

// Count returns the number of entries — O(1).
func (im *IndexedMap) Count() int { return len(im.Entries) }

// At returns the entry at index in canonical order — O(1). ok is false if out of
// range.
func (im *IndexedMap) At(index int) (e MapEntry, ok bool) {
	if index < 0 || index >= len(im.Entries) {
		return MapEntry{}, false
	}
	return im.Entries[index], true
}

// Get looks up key (an encoded key element) — O(log n). ok is false if absent.
func (im *IndexedMap) Get(key []byte) (value []byte, ok bool) {
	if i, found := im.Find(key); found {
		return im.Entries[i].Value, true
	}
	return nil, false
}

// Find returns the index of key in canonical order — O(log n). found is false if
// absent.
func (im *IndexedMap) Find(key []byte) (index int, found bool) {
	lo, hi := 0, len(im.Entries)
	for lo < hi {
		mid := lo + (hi-lo)/2
		switch bytesCompare(im.Entries[mid].Key, key) {
		case 0:
			return mid, true
		case -1:
			lo = mid + 1
		default:
			hi = mid
		}
	}
	return 0, false
}

// IndexedMapIterator yields entries of an IndexedMap in canonical order.
type IndexedMapIterator struct {
	entries []MapEntry
	i       int
}

// Next returns the next entry. ok is false at end.
func (it *IndexedMapIterator) Next() (e MapEntry, ok bool) {
	if it.i >= len(it.entries) {
		return MapEntry{}, false
	}
	e = it.entries[it.i]
	it.i++
	return e, true
}

// Iterator returns the entries in canonical (sorted) order.
func (im *IndexedMap) Iterator() *IndexedMapIterator {
	return &IndexedMapIterator{entries: im.Entries}
}

File: kotlin/src/Navigate.kt
/*
 * Navigation / query over an encoded struple buffer. Mirrors src/navigate.zig
 * and py/struple/_core.py (View / MapView / IndexedMap + Reader cursor).
 *
 * A buffer is a stream of elements; these helpers slice and inspect it without
 * decoding values. Results are sub-slices of the input, each itself a valid
 * struple buffer, so navigation composes and recurses.
 *
 * Stream ops (count/at/head/tail/nthRest/take) operate on the top-level element
 * stream; to descend into an array/map/set use containedItems to get its inner
 * stream, then view that.
 */
package struple

fun view(buf: ByteArray): View = View(buf)

class View(val buf: ByteArray) {

    fun reader(): Reader = Reader(buf)

    /** Number of top-level elements. */
    fun count(): Int {
        val r = reader()
        var n = 0
        while (r.skip()) n++
        return n
    }

    /** The element at [index] as a zero-copy sub-view, or null if out of range. */
    fun at(index: Int): ByteArray? {
        val r = reader()
        var i = 0
        while (true) {
            val v = r.nextView() ?: return null
            if (i == index) return v
            i++
        }
    }

    /** The first element (null if empty). */
    fun head(): ByteArray? = at(0)

    /** Everything after the first element (empty if 0 or 1 elements). */
    fun tail(): ByteArray {
        val r = reader()
        r.nextView()
        return r.rest()
    }

    /** Drop [n] elements; return the remaining stream. */
    fun nthRest(n: Int): ByteArray {
        val r = reader()
        var i = 0
        while (i < n) { if (!r.skip()) break; i++ }
        return r.rest()
    }

    /** The first [n] elements, as a contiguous sub-view. */
    fun take(n: Int): ByteArray {
        val r = reader()
        var i = 0
        while (i < n) { if (!r.skip()) break; i++ }
        return buf.copyOfRange(0, r.pos)
    }

    /** The type code of the first element (null if empty). */
    fun headType(): Int? = if (buf.isNotEmpty()) buf[0].toInt() and 0xFF else null

    fun isNil(): Boolean = headType() == Tc.NIL
    fun isUndefined(): Boolean = headType() == Tc.UNDEF
    fun isBool(): Boolean { val t = headType() ?: return false; return t == Tc.BOOL_FALSE || t == Tc.BOOL_TRUE }
    fun isInt(): Boolean {
        val t = headType() ?: return false
        return t == Tc.INT_ZERO || t == Tc.INT_NEG_BIG || t == Tc.INT_POS_BIG ||
            (t in 0x10..0x1F) || (t in 0x21..0x30)
    }
    fun isFloat(): Boolean { val t = headType(); return t == Tc.FLOAT32 || t == Tc.FLOAT64 }
    fun isDecimal(): Boolean = headType() == Tc.DECIMAL
    fun isNumber(): Boolean = isInt() || isFloat() || isDecimal()
    fun isTimestamp(): Boolean = headType() == Tc.TIMESTAMP
    fun isUuid(): Boolean = headType() == Tc.UUID
    fun isString(): Boolean = headType() == Tc.STRING
    fun isBytes(): Boolean = headType() == Tc.BYTES
    fun isArray(): Boolean = headType() == Tc.ARRAY
    fun isMap(): Boolean = headType() == Tc.MAP
    fun isSet(): Boolean = headType() == Tc.SET
    fun isContainer(): Boolean { val t = headType(); return t == Tc.ARRAY || t == Tc.MAP || t == Tc.SET }

    /**
     * The first element's framed body when it is a container (escapes intact,
     * zero-copy). Null if the head isn't a container. The Reader returns the
     * un-escaped inner stream, so for the framed body we slice it out of the raw
     * element bytes: [type code][escaped body][terminator].
     */
    fun containerBody(): ByteArray? {
        if (!isContainer()) return null
        val raw = reader().nextView() ?: return null
        return raw.copyOfRange(1, raw.size - 1)
    }

    /** The container's inner element stream, un-escaped (null if head isn't a container). */
    fun containedItems(): ByteArray? {
        if (!isContainer()) return null
        return when (val e = reader().next() ?: return null) {
            is Element.Arr -> e.inner
            is Element.MapElem -> e.inner
            is Element.SetElem -> e.inner
            else -> null
        }
    }
}

/**
 * Reads key/value pairs from a map's inner stream (the un-escaped body from
 * View.containedItems). Keys are canonical (sorted), so get early-exits.
 */
class MapView(val inner: ByteArray) {

    fun count(): Int = View(inner).count() / 2

    data class Entry(val key: ByteArray, val value: ByteArray) {
        override fun equals(other: Any?) =
            other is Entry && key.contentEquals(other.key) && value.contentEquals(other.value)
        override fun hashCode() = 31 * key.contentHashCode() + value.contentHashCode()
    }

    fun entries(): List<Entry> {
        val r = Reader(inner)
        val out = ArrayList<Entry>()
        while (true) {
            val k = r.nextView() ?: break
            val v = r.nextView() ?: throw StrupleException("struple: malformed map")
            out.add(Entry(k, v))
        }
        return out
    }

    /** Look up the value bytes for an encoded key element. Ordered scan, early exit. */
    fun get(key: ByteArray): ByteArray? {
        val r = Reader(inner)
        while (true) {
            val k = r.nextView() ?: return null
            val v = r.nextView() ?: throw StrupleException("struple: malformed map")
            when (order(k, key)) {
                0 -> return v
                1 -> return null // gt -> past it (canonical order)
            }
        }
    }

    /** Materialize a random-access index for O(log n) get and O(1) at. */
    fun indexed(): IndexedMap = IndexedMap(inner)
}

/**
 * A map's entries materialized into a random-access index. Building it is one
 * O(n) pass over the inner stream; thereafter get is an O(log n) binary search
 * (canonical key order means a key byte compare IS the sort order) and at is O(1).
 */
class IndexedMap(inner: ByteArray) {
    val entries: List<MapView.Entry>

    init {
        val list = ArrayList<MapView.Entry>()
        val r = Reader(inner)
        while (true) {
            val k = r.nextView() ?: break
            val v = r.nextView() ?: throw StrupleException("struple: malformed map")
            list.add(MapView.Entry(k, v))
        }
        entries = list
    }

    /** Number of entries — O(1). */
    fun count(): Int = entries.size

    /** The entry at [index] in canonical (sorted) order — O(1); null if out of range. */
    fun at(index: Int): MapView.Entry? = if (index in entries.indices) entries[index] else null

    /** The index of [key] in canonical order, or null — O(log n). */
    fun find(key: ByteArray): Int? {
        var lo = 0
        var hi = entries.size
        while (lo < hi) {
            val mid = lo + (hi - lo) / 2
            when (order(entries[mid].key, key)) {
                0 -> return mid
                -1 -> lo = mid + 1
                else -> hi = mid
            }
        }
        return null
    }

    /** Look up the value bytes for an encoded key — O(log n) binary search. */
    fun get(key: ByteArray): ByteArray? {
        val i = find(key) ?: return null
        return entries[i].value
    }

    /** Entries in canonical (sorted) order. */
    fun iterator(): Iterator<MapView.Entry> = entries.iterator()
}
